File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-3.cpp
class Solution {
   public:
    int findMin(vector<int>& nums) {
        const int numSize = static_cast<int>(nums.size());
        int L = 0;
        int R = numSize - 1;

        int result = nums[0];

        while (L <= R) {
            if (nums[L] < nums[R]) {
                result = min(result, nums[L]);
                break;
            }
            int M = (L + R) / 2;
            result = min(result, nums[M]);

            if (nums[M] >= nums[L]) {
                L = M + 1;
            } else {
                R = M - 1;
            }
        }

        return result;
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-3.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        result = max(nums)
        curr_max = 1
        curr_min = 1

        for n in nums:
            (curr_max, curr_min) = (
                max(n, n * curr_max, n * curr_min),
                min(n, n * curr_max, n * curr_min),
            )
            result = max(result, curr_max)
        return result
